geom/type_multi_point.go

```diff
@@ -26,39 +25,11 @@ func NewMultiPointFromPoints(pts []Point, opts ...ConstructorOption) MultiPoint
 		ctype &= p.CoordinatesType()
 	}
 
-	var empty BitSet
-	floats := make([]float64, 0, len(pts)*ctype.Dimension())
-	for i, pt := range pts {
-		c, ok := pt.Coordinates()
-		if !ok {
-			empty.Set(i, true)
-		}
-		floats = append(floats, c.X, c.Y)
-		if ctype.Is3D() {
-			floats = append(floats, c.Z)
-		}
-		if ctype.IsMeasured() {
-			floats = append(floats, c.M)
-		}
-	}
-	seq := NewSequence(floats, ctype)
-	return NewMultiPointWithEmptyMask(seq, empty, opts...)
-}
-
-// NewMultiPoint creates a new MultiPoint from a sequence of Coordinates.
-func NewMultiPoint(seq Sequence, opts ...ConstructorOption) MultiPoint {
-	return MultiPoint{seq, BitSet{}}
-}
-
-// NewMultiPointWithEmptyMask creates a new MultiPoint from a sequence of
-// coordinates. If there are any positions set in the BitSet, then these are
-// used to indicate that the corresponding point in the sequence is an empty
-// point.
-func NewMultiPointWithEmptyMask(seq Sequence, empty BitSet, opts ...ConstructorOption) MultiPoint {
-	return MultiPoint{
-		seq,
-		empty.Clone(), // clone so that the caller doesn't have access to the internal empty set
+	cp := make([]Point, len(pts))
+	for i := range cp {
+		cp[i] = pts[i].ForceCoordinatesType(ctype)
 	}
+	return MultiPoint{cp, ctype}
```

Comment:
this block of code is quite similar to `ForceCoordinatesType`; perhaps there's an opportunity to reduce duplication?

Reply:
That's a good catch.

The most obvious way to reduce the duplication is to just do `return MultiPoint{pts, ctype}.ForceCoordinatesType(ctype)`, but that has the awkward side effect that some of the invariant of the temporary `MultiPoint` don't hold (i.e. the `ctype` doesn't necessarily match that of each `Point` in the slice). Even though that invariant is only broken very temporarily, it makes me a bit uneasy.

So instead, in 9cb5049ff00ac71a41243317fa1ee0b062cebec9 I've split out the "force a slice of Points to have a particular ctype" logic into a new function, and am calling that from the `NewMultiPoint` function and the `ForceCoordinatesType` method.

Reply:
As an aside, the invariant wasn't explicitly documented anywhere. I've added documentation in 0b5e8bd247fbb9c0f441fa90fd18916555a8e731